lib/unicorn/http_server.rb

```diff
@@ -475,7 +478,10 @@ def murder_lazy_workers
       next_sleep = 0
       logger.error "worker=#{worker.nr} PID:#{wpid} timeout " \
                    "(#{diff}s > #{@timeout}s), killing"
-      before_murder.call(self, worker, wpid) if before_murder
+      if before_murder && !@before_murder_called.include?(wpid)
+        before_murder.call(self, worker, wpid)
+        @before_murder_called.add(wpid) # This ensures we call before_murder only once
+      end
```

Comment:
Could write this as:
```ruby
before_murder_call(self, worker, wpid) if before_murder and @before_murder_called.add?(wpid)
```
The `.add?` call adds the `wpid` to the `Set` and returns `nil` if it was already there or `self` if it wasn't.

Not sure if that's too much code :golf: